code/game/objects/items/devices/uplinks.dm

```diff
@@ -105,12 +105,25 @@ GLOBAL_LIST_EMPTY(world_uplinks)
 		to_chat(usr, "<span class='warning'>You have redeemed this discount already.</span>")
 		return
 	UI.buy(src,usr)
-	if(UI.limited_stock > 0) // only decrement it if it's actually limited
-		UI.limited_stock--
 	SStgui.update_uis(src)
 
 	return TRUE
 
+/obj/item/uplink/proc/mass_purchase(datum/uplink_item/UI, reference, quantity = 1)
+	// jamming check happens in ui_act
+	if(!UI)
+		return
+	if(quantity <= 0)
+		return
+	if(UI.limited_stock == 0)
+		return
+	if(UI.limited_stock > 0 && UI.limited_stock < quantity)
+		quantity = UI.limited_stock
+	var/list/bought_things = list()
+	for(var/i in 1 to quantity)
+		bought_things += UI.buy(src, usr, put_in_hands = FALSE)
```

Comment:
This can contain null elements you'll have to filter out with the suggestions. If it fails, then you might also just as well stop trying since the coming tries will also fail.

Reply:
it will now break if it returns nothing